---
title: "Load screen data"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
theme_set(theme_bw())
knitr::opts_chunk$set(echo = TRUE)
```

```{r }

loadAndCollapse <- function(file) {
  counts <- read.csv(paste0("./barcode_counts/", file), header = T, stringsAsFactors = FALSE)
  counts <- counts[counts$barcode != "no_match", ]
  fcounts <- as.matrix(counts [grep("\\.1$", names(counts))]) # forward
  rcounts <- as.matrix(counts [grep("\\.2$", names(counts))]) # reverse
  alltcounts <- fcounts + rcounts
  colnames(alltcounts) <- gsub("\\.1", "", colnames(alltcounts))
  rownames(alltcounts) <- counts$gene
  alltcounts <- as.data.frame(t(alltcounts))
  alltcounts$Index <- 1:nrow(alltcounts)
  alltcounts$file <- file
  row.names(alltcounts) <- NULL
  return(alltcounts)
}


metadata <- read_csv("./metadata/SampleAssignments.csv")
files <- unique(metadata$Run)
files <- files[!is.na(files)]
files <- paste0(files, ".csv")
dfs <- lapply(files, loadAndCollapse)
abundances <- plyr::rbind.fill(dfs) %>%
  mutate(Run = gsub(".csv", "", file))

abundances <- inner_join(abundances, metadata, by = c("Index", "Run"), how = "left")

abundances <- abundances %>% filter(Condition != "DoNotAnalyse")

abundances <- abundances %>% pivot_longer(cols = c(starts_with("PBANKA"),`PC_API0037`,`p230p-tag` ), names_to = "gene", values_to = "count")

# Set previously implicit 0 values for barcodes not found.
abundances <- abundances %>% replace_na(list(count = 0))

# Calculate barcode proportions per sample
abundances <- abundances %>%
  group_by(Pool, Condition, Mouse, Replicate) %>%
  mutate(proportion = count / sum(count))

# If this barcode isn't really in the pool at all, we shouldn't analyse it. Here we identify barcodes at least sometimes present at greater than 0.01% on a per-pool basis:
wellrepresentedinpassage <- abundances %>%
  filter(Condition == "control" || Condition == "Negsort") %>%
  group_by(gene, Pool) %>%
  filter(max(proportion) > 0.0001) %>%
  summarise(toinclude = T)

# Filter in only these cases
abundances<-abundances %>% left_join(wellrepresentedinpassage) %>% filter(toinclude==T)

# Recalculate proportions, adding 0.5 to each count to make logarithmic calculations possible by avoiding zero (this is reasonable since these values might in fact be 0.4 reads)
abundances <- abundances %>% mutate(count=count+0.5)

abundances<-abundances %>% group_by(Pool,Condition,Mouse,Replicate) %>% mutate(proportion=(count)/sum(count))

#We'll use log-2 proportions for most downsread calculations
abundances<- abundances %>% mutate(logprop=log2(proportion))

abundances <- abundances %>% mutate(Mouse = as.character(case_when(Mouse=='A' ~ 1,Mouse=='B' ~ 2,Mouse=='C' ~ 3,TRUE ~ as.numeric(Mouse))))
```

Diagnostic plot to verify reasonable coverage in each sample, grouped by sequencing run.
```{r}
totals<-abundances%>% group_by(Pool,Condition,Mouse,Replicate,Run)%>% summarise(sum=sum(count))

ggplot(totals,aes(x=paste(Pool,Condition,Mouse,Replicate),fill=Pool,y=sum))+geom_bar(stat="identity")+facet_wrap(~Run,scales="free",ncol=3) +theme_bw()+ theme(axis.text.x = element_text(angle = 90, hjust = 1,size=5))


```

Diagnostic plot of reproducibility of technical replicates:

```{r}

for_scatter <- abundances %>% select(Pool,Condition,Mouse,Replicate,gene,logprop) %>% pivot_wider(names_from="Replicate",values_from="logprop",names_prefix ="replicate")

ggplot(for_scatter,aes(x=replicate1,y=replicate2,color=Pool))+facet_wrap(~paste(Pool,Mouse,Condition))+geom_point(size=0.01)
```

Plot correlations. (Sorts 2 and 4 represented a sort of low intensity marker genes to investigate their potential and were not used further.) Negsort represents a sort of GFP- and RFP- negative parasites (DNA positive) whilst control represents a sample from the Histodenz gradient.
```{r,fig.height=30,fig.width=10}
library(zoo)

variance <- abundances %>%
  group_by(Pool, Condition, Mouse, gene) %>%
  summarise(sd = sd(logprop), val1 = logprop[1], val2 = logprop[2], meanlogprop = mean(logprop))

cors <- variance %>%
  group_by(Condition, Mouse, Pool) %>%
  summarise(cor = cor(val1, val2), n = n())

ggplot(cors, aes(x=Condition,y=cor,color=Mouse))+facet_wrap(~Pool)+geom_point() + theme(axis.text.x = element_text(angle = 90))


variance <- variance %>%
  group_by(Pool, Condition, Mouse) %>%
  arrange(-meanlogprop) %>%
  mutate(medsd = rollmean(x = sd, k = 11, fill = "extend"))
variance <- variance %>% mutate(monotonicmedsd = cummax(medsd))


ggplot(variance, aes(x=meanlogprop,y=sd,color=Pool))+facet_grid(Mouse+Condition~Pool)+geom_point() + theme(axis.text.x = element_text(angle = 90))+geom_smooth(color="gray") + geom_line(aes(y=monotonicmedsd), color="black")


```

For some experiments a negative sorted population was not collected. Instead we use as a negative control the Nycodenz-purified parasite pellet
```{r}
variance <- variance %>%
  ungroup() %>%
  mutate(Condition = case_when(Pool=="SP3" & Condition=="control" ~ "Control",
                               Pool!="SP3" & Condition=="Negsort" ~ "Control",
                               TRUE ~ Condition
                               )
           )  %>% filter(Condition %in% c("RFPsort","GFPsort","Control"))
ggplot(variance, aes(x=meanlogprop,y=sd,color=Pool))+facet_grid(Mouse+Condition~Pool)+geom_point() + theme(axis.text.x = element_text(angle = 90))+geom_smooth(color="gray") + geom_line(aes(y=monotonicmedsd), color="black")


noncontrol <- variance %>% filter(Condition != "Control")
control <- variance %>% filter(Condition == "Control")

comparison <- inner_join(ungroup(control), ungroup(noncontrol), by = c("Mouse", "gene", "Pool") , suffix=c("Control","Other"))


```


```{r}
gaussianMeanAndVariance <- function(vals, variances, return) {
  df <- data.frame(value = vals, variance = variances)
  df <- df[complete.cases(df), ]

  vals <- df$value
  variances <- df$variance
  if (length(vals) == 1) {
    var <- variances[1]
    mean <- vals[1]
  }
  else {
    precs <- 1 / variances

    mean <- sum(vals * precs) / sum(precs)
    var1 <- (1 / sum(precs)) * (1 / (length(vals) - 1)) * sum((vals - mean)**2 / variances)
    var2 <- 1 / sum(precs)
    if (is.na(var1)) {
      var1 <- 0
    }
    var <- max(var1, var2)
  }
  return(list("mean"=mean,"variance"=variance))
}
```


Calculate the difference in log2 proportions and propagate its uncertainty through
```{r}
comparison <- comparison %>% mutate(difference = meanlogpropOther-meanlogpropControl)
comparison <- comparison %>% mutate(differencesd = sqrt(monotonicmedsdOther^2 + monotonicmedsdControl^2))
comparison <- comparison %>% mutate(Condition = ConditionOther)

```


```{r}
comparisonmerge <- comparison


comparisonmerge <- comparisonmerge %>% mutate(p = 1 - pnorm(0, mean = difference, sd = differencesd))
comparisonmerge <- comparisonmerge %>% mutate(diffmax = difference + 2 * differencesd)
comparisonmerge <- comparisonmerge %>% mutate(diffmin = difference - 2 * differencesd)
comparisonmerge <- comparisonmerge %>% mutate(power = ifelse(diffmin > -1, "notreduced", ifelse(diffmax < (-1), "reduced", "nopower")))
GFP <- filter(comparisonmerge, Condition == "GFPsort")
GFP$gene <- as.character(GFP$gene)
RFP <- filter(comparisonmerge, Condition == "RFPsort")
RFP$gene <- as.character(RFP$gene)
# joint=full_join(GFP,RFP,by=c("gene","Pool"))
joint <- full_join(GFP, RFP, by = c("gene"), suffix = c(".gfp", ".rfp"))
joint$minmax <- pmin(joint$diffmax.gfp, joint$diffmax.rfp)
joint <- mutate(joint,
  class = case_when(
    diffmax.gfp < -1 & diffmax.rfp < -1 ~ "lossOfBoth",
    diffmax.gfp < -1 ~ "lossOfMales",
    diffmax.rfp < -1 ~ "lossOfFemales",
    diffmin.rfp > 1 & diffmin.gfp > 1 ~ "gainOfBoth",
    diffmin.rfp > 1 ~ "gainOfFemales",
    diffmin.gfp > 1 ~ "gainOfMales",
    diffmax.gfp > -1 & diffmax.rfp > -1 ~ "No change",
    TRUE ~ "Unsure"
  )
)
```




```{r}
library(tidyverse)
counts <- read_tsv("./ap2gtimecourse/counts.txt", col_names = T)

phenodata <- read_tsv("./ap2gtimecourse/phenodata.txt", col_names = T)
phenodata$X1 <- gsub("X820", "820", phenodata$X1)

narrow <- counts %>% gather("sample", "count", -X1)
colnames(narrow) <- c("gene", "X1", "count")
narrow <- narrow %>% filter(count > 0)

both <- inner_join(phenodata, narrow)
both <- both %>%
  group_by(X1) %>%
  mutate(proportion = count / sum(count))


counts2 <- as.matrix(select(counts, -X1))
rownames(counts2) <- counts$X1


timepoints <- unique(phenodata$time)
i <- 9
cols <- colnames(counts2)
library(DESeq2)

performDE <- function(filter) {
  abc <- tibble()

  filter2 <- filter %>% filter(Strain == "AP2Goe")
  print(unique(filter$time))
  if (length(unique(filter2$rapamycin)) > 1) {
    
    
    counttable <- counts2[, base::match(filter2$X1, cols)]
    dds <- DESeqDataSetFromMatrix(countData = counttable,
                              colData = filter2,
                              design= ~  rapamycin)
    dds <- DESeq(dds)
    

    res <- results(dds)
    abc <- as.tibble(res) 
    abc$id = rownames(results(dds))
    abc$timepoint <- unique(filter$time)
  }
  return(abc)
}


new <- phenodata %>%
  group_by(time) %>%
  filter(time != "3h") %>%
  do(performDE(.))


oneversion <- joint
asexual <- read_csv("./otherdata/asexualscreendata.csv", col_names = T)
merge <- right_join(oneversion, asexual)
new$id <- gsub(".1$", "", new$id)
new$time2 <- as.numeric(gsub("h", "", new$time))
merge2 <- inner_join(new, merge, by = c("id" = "current_version_ID"))
filt <- merge2 %>%
  group_by(timepoint) %>%
  arrange(pvalue) %>%
  filter(row_number() < 50) %>%
  mutate(n = row_number())
ggplot(filt, aes(y = -n, x = time2, fill = class)) +
  geom_tile() +
  geom_line(data = filter(filt, !is.na(class)), aes(group = id, color = class, alpha = 0.5))


filt <- merge2 %>%
  group_by(timepoint) %>%
  arrange(pvalue) %>%
  mutate(n = row_number())

ggplot(filt %>% filter(class !="No change", n<50), aes(y = n, x = time2, fill = class,color=class, group=id)) + geom_point(data=filt %>% filter(class =="No change", n<50) ,color="gray",alpha=0.2) + geom_line(data=filt %>% filter(class =="No change", n<50) ,color="gray",alpha=0.2)+
  geom_point() +geom_line()+theme_bw()+
  labs(y = "Genes ranked at each timepoint by p-value", x = "Timepoint (hours) after AP2-G induction")
```